### PR TITLE
Create the AST type definitions.

### PR DESCRIPTION
**Pull request by kontheocharis**
This is the frontend-agnostic AST that we will use to perform typechecking and code emitting.
Loosely modelled after the Haskell version.

### CODE REVIEW

File: src/ast.rs [hunk 1]
```diff
@@ -0,0 +1,547 @@
+//! Frontend-agnostic Hash abstract syntax tree type definitions.
+#![allow(dead_code)]
```
Comment (feds01) [inline src/ast.rs:2]: This is temporary?

File: src/ast.rs [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+    /// A block.
+    Block(AstNode<Block>),
+    /// An `import` call.
+    Import(AstNode<ImportPath>),
+}
+
+/// A module.
+///
+/// Represents a parsed `.hash` file.
+#[derive(Debug, Clone)]
```
Comment (feds01) [inline src/ast.rs:543]: What does deriving `Debug` and `Clone` do to a particular struct?

File: src/main.rs [hunk 3]
```diff
@@ -1,3 +1,6 @@
+mod ast;
```
Comment (feds01) [inline src/main.rs:1]: I guess we can re-organise the sources a little later once we finish up all the stuff with parsing.

**Review by feds01 — APPROVED** [review]
Looks really good, should we add copyright banners to the files?

File: src/ast.rs (hunk 1 above)
Comment (kontheocharis) [reply src/ast.rs:2]: Yes, cause otherwise it will complain that we aren't using any of the items.

File: src/ast.rs (hunk 2 above)
Comment (kontheocharis) [reply src/ast.rs:543]: https://doc.rust-lang.org/rust-by-example/trait/derive.html

File: src/main.rs (hunk 3 above)
Comment (kontheocharis) [reply src/main.rs:1]: Yeah we probably wanna create separate crates for some stuff.

**Review by feds01 — APPROVED** [review]

**Comment by feds01** [comment]
Looks good, just got to resolve the merge conflicts and then we can merge it into main

**Pull request merged**